src/main/java/com/aws/greengrass/mqtt/bridge/clients/LocalMqtt5Client.java

```diff
@@ -0,0 +1,391 @@
+            topicsToSubscribe.removeAll(subscribedLocalMqttTopics);
+
+            LOGGER.atDebug().kv("topics", topicsToSubscribe).log("Subscribing to MQTT topics");
+
+            subscribeToTopics(topicsToSubscribe);
+
+        }
+    }
+
+    private void subscribe(String topic) {
+        if (client.getIsConnected()) {
+            SubscribePacket subscribePacket = new SubscribePacket.SubscribePacketBuilder()
+                    .withSubscription(topic, QOS).build();
+            LOGGER.atDebug().kv(TOPIC, topic).log("Subscribing to MQTT topic");
+            //TODO: deal with subscribeFuture
+            this.subscribeFuture = client.subscribe(subscribePacket);
+            try {
+                SubAckPacket subAckPacket = this.subscribeFuture.get();
+                //TODO: Fix logging
+                LOGGER.atDebug().kv("reason", subAckPacket.getReasonString()).log("while subscribing");
+            } catch (InterruptedException | ExecutionException e) {
+                LOGGER.atDebug().setCause(e).kv(TOPIC, topic).log("failed to subscribe");
+            }
+
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void subscribeToTopics(Set<String> topics) {
+        //TODO: Think about making this async (non-blocking)
+
+        // retry until interrupted
+        topics.forEach(s -> {
+            try {
+                RetryUtils.runWithRetry(mqttExceptionRetryConfig, () -> {
+                    subscribe(s);
+                    // useless return
+                    return null;
+                }, "subscribe-mqtt5-topic", LOGGER);
+                subscribedLocalMqttTopics.add(s);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                //TODO: Subscribe doesn't throw an exception
+                LOGGER.atError().setCause(e).kv(TOPIC, s).log("Failed to subscribe");
+            }
+        });
+    }
+
+    private void resubscribe(boolean sessionPresent) {
+        updateSubscriptionsInternal(sessionPresent);
+//        subscribedLocalMqttTopics.clear();
+//        updateSubscriptionsInternal();
+    }
+
+    private void unsubscribe(String topic) {
+        if (client.getIsConnected()) {
+            UnsubscribePacket unsubscribePacket =
+                    new UnsubscribePacket.UnsubscribePacketBuilder().withSubscription(topic).build();
+            LOGGER.atDebug().kv(TOPIC, topic).log("Unsubscribing from MQTT topic");
+            CompletableFuture<UnsubAckPacket> unsubscribeFuture = client.unsubscribe(unsubscribePacket);
+            try {
+                UnsubAckPacket unsubAckPacket = unsubscribeFuture.get();
+                //TODO: Fix logging
+                LOGGER.atDebug().kv("reason", unsubAckPacket.getReasonString()).log("while unsubscribing");
+            } catch (InterruptedException | ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)